### PR TITLE
ARROW-16852: [C++] Migrate remaining kernels to use ExecSpan, remove ExecBatchIterator

### PR DESCRIPTION
**Pull request by wesm**
This completes the porting to use ExecSpan everywhere. I also changed the ExecBatchIterator benchmarks to use ExecSpan to show the performance improvement in input splitting that we've talked about in the past:

Splitting inputs into small ExecSpan:

```
------------------------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------
BM_ExecSpanIterator/1024      205671 ns       205667 ns         3395 items_per_second=4.86223k/s
BM_ExecSpanIterator/4096       54749 ns        54750 ns        13121 items_per_second=18.265k/s
BM_ExecSpanIterator/16384      15979 ns        15979 ns        42628 items_per_second=62.5824k/s
BM_ExecSpanIterator/65536       5597 ns         5597 ns       125099 items_per_second=178.668k/s
```

Splitting inputs into small ExecBatch:

```
-------------------------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------
BM_ExecBatchIterator/1024    17163432 ns     17163171 ns           41 items_per_second=58.2643/s
BM_ExecBatchIterator/4096     4243467 ns      4243316 ns          163 items_per_second=235.665/s
BM_ExecBatchIterator/16384    1093680 ns      1093638 ns          620 items_per_second=914.38/s
BM_ExecBatchIterator/65536     272451 ns       272435 ns         2584 items_per_second=3.6706k/s
```

Because the input in this benchmark has 1M elements, this shows that splitting into 1024 chunks of size 1024 adds only 0.2ms of overhead with ExecSpanIterator versus 17.16ms of overhead with ExecBatchIterator (> 80x improvement). 

This won't by itself do much to impact performance in Acero but things for the community to explore in the future are the following (this work that I've been doing has been a precondition to consider this):

* A leaner ExecuteScalarExpression implementation that reuses temporary allocations (ARROW-16758)
* Parallel expression evaluation
* Better defining morsel (~1M elements) versus task (~1K elements) granularity in execution 
* Work stealing so that we don't "hog" the thread pools, and we keep the work pinned to a particular CPU core if there are other things going on at the same time

### CODE REVIEW

**Comment by wesm** [comment]
The only CI failures are the linkage issues related to protobuf/otel

File: cpp/src/arrow/compute/kernels/aggregate_basic_internal.h [hunk 1]
```diff
@@ -448,9 +449,12 @@ struct MinMaxImpl : public ScalarAggregator {
     return Status::OK();
   }
 
-  Status ConsumeArray(const ArrayType& arr) {
+  Status ConsumeArray(const ArraySpan& arr_span) {
     StateType local;
 
+    // TODO(wesm): do not use ToArrayData
+    ArrayType arr(arr_span.ToArrayData());
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:456]: Is this something that you plan to do in a later PR? Or can it just be done here?

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 2]
```diff
@@ -471,45 +471,6 @@ struct ExactQuantiler<InType, enable_if_t<is_decimal_type<InType>::value>> {
...
-    }
-  } else {
-    double* out_buffer = output->template GetMutableValues<double>(1);
-    for (int64_t i = 0; i < output->length; i++) {
-      out_buffer[i] = DataPointToDouble(UnboxScalar<T>::Unbox(scalar), *scalar.type);
-    }
-  }
-  return Status::OK();
-}
-
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_quantile.cc:512]: Neat removal :-)

File: cpp/src/arrow/compute/kernels/hash_aggregate_test.cc [hunk 3]
```diff
@@ -202,14 +202,16 @@ Result<Datum> GroupByUsingExecPlan(const std::vector<Datum>& arguments,
...
   BatchesWithSchema input;
   input.schema = schema(std::move(scan_fields));
-  ExecBatch batch;
-  while (batch_iterator->Next(&batch)) {
-    if (batch.length == 0) continue;
-    input.batches.push_back(std::move(batch));
+  ExecSpan span;
+  while (span_iterator.Next(&span)) {
+    if (span.length == 0) continue;
+    // TODO(wesm): investigate possibility of ExecPlans using ExecSpan
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/hash_aggregate_test.cc:213]: Should this be a JIRA instead? TODOs sprinkled in code are often ignored.

**Review by pitrou — COMMENTED** [review]
Very nice, just a couple minor comments.

File: cpp/src/arrow/compute/exec_internal.h [hunk 4]
```diff
@@ -84,7 +51,10 @@ class ARROW_EXPORT ExecSpanIterator {
   /// \param[in] batch the input ExecBatch
   /// \param[in] max_chunksize the maximum length of each ExecSpan. Depending
   /// on the chunk layout of ChunkedArray.
-  Status Init(const ExecBatch& batch, int64_t max_chunksize = kDefaultMaxChunksize);
+  /// \param[in] promote_if_all_scalars if all of the values are scalars,
+  /// return them in each ExecSpan as ArraySpan of length 1
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/compute/exec_internal.h:55]: I'm adding a comment that it should be set to true for scalar/vector executors and false for scalar/hash aggregation execution

File: cpp/src/arrow/compute/exec_internal.h [hunk 5]
```diff
@@ -110,6 +80,7 @@ class ARROW_EXPORT ExecSpanIterator {
   bool initialized_ = false;
   bool have_chunked_arrays_ = false;
   bool have_all_scalars_ = false;
+  bool promote_if_all_scalars_ = false;
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/compute/exec_internal.h:83]: +1

File: cpp/src/arrow/compute/kernels/hash_aggregate_test.cc (hunk 3 above)
Comment (wesm) [reply cpp/src/arrow/compute/kernels/hash_aggregate_test.cc:213]: Removing this comment -- this is something we can tackle in more general refactoring of the execution engine (to introduce a distinction between tasks and morsels) in the future

File: cpp/src/arrow/compute/row/grouper.cc [hunk 6]
```diff
@@ -294,20 +282,21 @@ struct GrouperFastImpl : Grouper {
...
+  Result<Datum> ConsumeImpl(const ExecSpan& batch) {
     int64_t num_rows = batch.length;
     int num_columns = batch.num_values();
     // Process dictionaries
     for (int icol = 0; icol < num_columns; ++icol) {
       if (key_types_[icol].id() == Type::DICTIONARY) {
-        auto data = batch[icol].array();
-        auto dict = MakeArray(data->dictionary);
+        const ArraySpan& data = batch[icol].array;
+        // TODO(wesm): do not require ToArrayData here
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/compute/row/grouper.cc]: Removed the comment for now, each of the usages of `ToArrayData` can be given consideration in the future

**Comment by wesm** [comment]
Thanks for the review. Will merge when CI green

**Review by lidavidm — APPROVED** [review]

**Comment by wesm** [comment]
Merging. The CI issues look like assorted flakiness to me

**Pull request merged**